KIPs/kip-113.md

```diff
@@ -0,0 +1,252 @@
+6. Bytecode-injection: inject a bytecode to a predefined address at hardfork.
+
+We can view each method from the following point of view:
+
+##### Explicit contract logic upgrade
+
+| Methods                                                | [1] HeaderGov | [2] ContractGov                        | [3] SpareContract | [4] ChainConfig | [5] Precompiled-contract | [6] Bytecode-injection |
+| ------------------------------------------------------ | ------------- | -------------------------------------- | ----------------- | --------------- | ------------------------ | ---------------------- |
+| Upgrade with hardfork                                  | X             | X                                      | X                 | O               | O                        | O                      |
+| Managing address change timing                         | O             | O                                      | X                 | O               | O                        | O                      |
+| Awareness of subscriber contracts after address update | X             | △<br> Latest GovParam address required | O                 | X               | O                        | X                      |
+
+In terms of explicit contract logic upgrade, ChainConfig, Precompiled-contract, and Bytecode-injection are recommended.
+
+##### Usability from ServiceChain
+
+| Methods                      | [1] HeaderGov  | [2] ContractGov           | [3] SpareContract              | [4] ChainConfig | [5] Precompiled-contract          | [6] Bytecode-injection |
+| ---------------------------- | -------------- | ------------------------- | ------------------------------ | --------------- | --------------------------------- | ---------------------- |
+| Binary modification required | O              | O                         | X<br>Patch AddressBook address | O               | X<br> Patch the hardcoded address | O                      |
+| Contract deployment required | X<br> Registry | X <br> Registry, GovParam | X <br> Registry, AddressBook   | X<br> Registry  | X<br> Registry                    | X<br> AddressBook      |
+
+In terms of usability from ServiceChain, HeaderGov, ContractGov, ChainConfig, and Bytecode-injection are recommended.
+
+##### Others
+
+| Methods                                            | [1] HeaderGov                 | [2] ContractGov               | [3] SpareContract             | [4] ChainConfig      | [5] Precompiled-contract | [6] Bytecode-injection |
+| -------------------------------------------------- | ----------------------------- | ----------------------------- | ----------------------------- | -------------------- | ------------------------ | ---------------------- |
+| Immediate panic at hardfork from out-of-date nodes | △<br> At GovernanceData block | X<br> At the activation block | △<br> At SpareContract update | O                    | O                        | O                      |
```

Comment:
hmm immediate panic is a good thing?

Reply:
I think the sooner you find out you didn't upgrade the node, the better. If it panics later, there can be situations where you have to rollback to the hardfork (e.g., incompatible DB).

Reply:
I got your point, but it would be good to rephrase it.
```suggestion
| Possible to alert immediately for out-of-date nodes | △<br> At GovernanceData block | X<br> At the activation block | △<br> At SpareContract update | O                    | O                        | O                      |
```

Reply:
you mean `immediately at hardfork` right?

Reply:
do you mean alert as in notify/warn?

Reply:
I think "alerting out-of-date nodes" is ambiguous; who alerts to whom and how?

Reply:
okay then, can you suggest something else? 

1. I want the statement to be more yes/no question.
2. what to emphasize is that whether we have a method to force them update the node. Otherwise, the node will not work. 

So, my another suggestion would be

"Possible to panic the node if out-dated"

Reply:
Initially the answers are
- O if panics at hardfork block,
- △ if panics later at a predictable block,
- X if panics at undetermined block.

If the question is "Possible to panic outdated nodes", they they're all yes. (they will panic some day)
To keep the answeres, I'll (1) change the question to "Possible to panic outdated nodes at hardfork block" and (2) change answers from △ to X.

Reply:
Sorry for the late reply. So, in this case, you think panicing at the hard fork block is the best one, dont't you?

Reply:
yes, i do think.

I removed this section, and we can further discuss in the upcoming KIP.